resolver-dns/src/main/java/io/netty/resolver/dns/DnsQueryContext.java

```diff
@@ -213,12 +213,13 @@ private void setFailure(String message, Throwable cause) {
            .append(" (no stack trace available)");
 
         final DnsNameResolverException e;
-        if (cause != null) {
-            e = new DnsNameResolverException(nameServerAddr, question(), buf.toString(), cause);
+        if (cause == null) {
+            // This was caused by an timeout so use DnsNameResolverTimeoutException to allow the user to
```

Comment:
IIUC this may not always be true. This method is also called when a write failure occurs (from `onQueryWriteCompletion`), right? Consider passing this exception (or some indication as to what the exception is) in to this method where the root cause is known instead of trying to infer it here.

Reply:
Yes and in this case `cause != null`. it is only null if we call it because of a timeout

Reply:
> it is only null if we call it because of a timeout

IIUC here is the stack trace that may call this method in a non-timeout case:

`DnsQueryContext#setFailure`
`DnsNameResolverContext#onQueryWriteCompletion`
`DnsNameResolverContext#writeQuery`

This may be the result of a transport error, not a timeout, no?

Reply:
@Scottmitch but only the one for the timeout will use `setFailure` and passing `null` into it as `cause`. So whenever `cause == null` we know its a timeout. I am missing something ?

Reply:
ahh my mistake. yes what you are saying makes sense. lgtm.